bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java

```diff
@@ -0,0 +1,272 @@
+
+        return value;
+    }
+
+    /**
+     * Write out any outstanding data
+     */
+    public void commitDatabase() {
+        String s = mapper.toJson(map);
+
+        synchronized (map) {
+            // Rename the file for backup
+            File rename = new File(file.getParent() + File.separator + "backup",
+                    System.currentTimeMillis() + "-" + file.getName());
+            file.renameTo(rename);
+
+            FileOutputStream outputStream;
+            try {
+                outputStream = new FileOutputStream(file, false);
+                outputStream.write(s.getBytes());
+                outputStream.close();
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+
+            deferredSince = 0;
+        }
+    }
+
+    private class CommitTimerTask extends TimerTask {
+        @Override
+        public void run() {
+            // Save the database
+            commitDatabase();
+
+            // Delete old backups
+            List<Long> fileTimes = new ArrayList<Long>();
+            File folder = new File(file.getParent() + File.separator + "backup");
+            File[] files = folder.listFiles();
+
+            // Get an array of file times from the filename
+            int count = files.length;
+            for (int i = 0; i < count; i++) {
+                if (files[i].isFile()) {
+                    String[] parts = files[i].getName().split(SEPARATOR);
+                    if (parts.length != 2 || !parts[1].equals(file.getName())) {
+                        continue;
+                    }
+                    long time = Long.parseLong(parts[0]);
+                    fileTimes.add(time);
+                }
+            }
+
+            // Sort, and delete the oldest
+            Collections.sort(fileTimes);
+            if (fileTimes.size() > MAX_FILES) {
+                for (int counter = 0; counter < fileTimes.size() - MAX_FILES; counter++) {
+                    File deleter = new File(file.getParent() + File.separator + "backup",
+                            fileTimes.get(counter) + SEPARATOR + file.getName());
+                    deleter.delete();
+                }
+            }
+        }
+    }
+
+    public synchronized void deferredCommit() {
+        // Handle a maximum time for deferring the commit.
+        // This stops a pathological loop preventing saving
+        if (deferredSince != 0 && deferredSince < System.currentTimeMillis() - MAX_DEFERRED_PERIOD) {
+            commitDatabase();
+        }
+        if (deferredSince == 0) {
+            deferredSince = System.currentTimeMillis();
+        }
+
+        // Stop any existing timer
+        if (commitTimerTask != null) {
+            commitTimerTask.cancel();
+            commitTimerTask = null;
+        }
+
+        // Create the timer task
+        commitTimerTask = new CommitTimerTask();
+
+        // Start the timer
+        commitTimer.schedule(commitTimerTask, WRITE_DELAY);
```

Comment:
Could WRITE_DELAY and MAX_DEFERRED_PERIOD be made user configurable so users can decrease write cycles to flash memory?